File: src/providers.ts
import { ButterclawConfig } from "./config.js";
import { trimTrailingSlash } from "./util.js";

export interface Message {
  role: "system" | "user" | "assistant";
  content: string;
}

export interface ProviderResponse {
  content: string;
  promptTokens?: number;
  completionTokens?: number;
  raw?: unknown;
}

export interface Provider {
  complete(messages: Message[]): Promise<ProviderResponse>;
}

export class ProviderError extends Error {}

export function estimateTokens(text: string): number {
  return Math.max(1, Math.ceil(text.length / 4));
}

export function buildProvider(config: ButterclawConfig): Provider {
  const candidates = buildProviderCandidates(config);
  if (candidates.length > 1) {
    return new FallbackProvider(candidates);
  }
  return candidates[0].provider;
}

export function buildProviderCandidates(config: ButterclawConfig): Array<{ label: string; provider: Provider }> {
  return [config, ...config.modelFallbacks.map((fallback) => fallbackConfig(config, fallback))].map((candidate) => ({
    label: `${candidate.provider}/${candidate.model}`,
    provider: buildSingleProvider(candidate)
  }));
}

function buildSingleProvider(config: ButterclawConfig): Provider {
  switch (config.provider) {
    case "mock":
      return new MockProvider(config.model);
    case "ollama":
      return new OllamaProvider(config.model, config.baseUrl ?? "http://localhost:11434", config.requestTimeoutSeconds);
    case "openai-compatible":
      return new OpenAICompatibleProvider(
        config.model,
        config.baseUrl ?? "https://openrouter.ai/api/v1",
        process.env[config.apiKeyEnv] ?? "",
        config.requestTimeoutSeconds
      );
    default:
      throw new ProviderError(`Unknown provider: ${config.provider}`);
  }
}

class FallbackProvider implements Provider {
  constructor(private readonly candidates: Array<{ label: string; provider: Provider }>) {}

  async complete(messages: Message[]): Promise<ProviderResponse> {
    const errors: string[] = [];
    for (const [index, candidate] of this.candidates.entries()) {
      try {
        const response = await candidate.provider.complete(messages);
        return index === 0
          ? response
          : {
              ...response,
              raw: { ...(typeof response.raw === "object" && response.raw ? response.raw : {}), fallbackUsed: candidate.label, fallbackErrors: errors }
            };
      } catch (error) {
        errors.push(`${candidate.label}: ${error instanceof Error ? error.message : String(error)}`);
      }
    }
    throw new ProviderError(`All model candidates failed:\n${errors.join("\n")}`);
  }
}

function fallbackConfig(base: ButterclawConfig, spec: string): ButterclawConfig {
  const [prefix, ...rest] = spec.split("/");
  const maybeProvider = prefix as ButterclawConfig["provider"];
  if (["mock", "ollama", "openai-compatible"].includes(maybeProvider) && rest.length) {
    return { ...base, provider: maybeProvider, model: rest.join("/") };
  }
  return { ...base, model: spec };
}

export class MockProvider implements Provider {
  constructor(private readonly model = "mock-local") {}

  async complete(messages: Message[]): Promise<ProviderResponse> {
    if (this.model.includes("fail")) {
      throw new ProviderError(`Mock provider forced failure for ${this.model}`);
    }
    const lastMessage = messages[messages.length - 1]?.content ?? "";
    const last = lastMessage.toLowerCase();
    const systemPrompt = messages.find((message) => message.role === "system")?.content ?? "";
    const canDelegate = systemPrompt.includes("- delegate_task:");
    if (last.includes("tool result for")) {
      return { content: mockToolResultAnswer(lastMessage) };
    }
    const requestedAgent = last.match(/\bask\s+([a-z0-9_.-]+)\b/);
    if (canDelegate && requestedAgent && last.includes("list")) {
      return {
        content: JSON.stringify({
          tool: "delegate_task",
          args: {
            role: requestedAgent[1],
            task: "list the files in this workspace"
          }
        })
      };
    }
    if (canDelegate && (last.includes("delegate") || last.includes("sub-agent") || last.includes("sub agent")) && last.includes("list")) {
      return { content: '{"tool":"delegate_task","args":{"role":"scout","task":"list the files in this workspace"}}' };
    }
    if (last.includes("list") && (last.includes("file") || last.includes("workspace"))) {
      return { content: '{"tool":"list_dir","args":{"path":"."}}' };
    }
    if (last.includes("read") && last.includes("readme")) {
      return { content: '{"tool":"read_file","args":{"path":"README.md"}}' };
    }
    return {
      content:
        "Butterclaw mock provider is running. Switch to --provider ollama or --provider openai-compatible for model-backed reasoning."
    };
  }
}

export class OpenAICompatibleProvider implements Provider {
  constructor(
    private readonly model: string,
    private readonly baseUrl: string,
    private readonly apiKey: string,
    private readonly timeoutSeconds: number
  ) {}

  async complete(messages: Message[]): Promise<ProviderResponse> {
    if (!this.apiKey) {
      throw new ProviderError("Missing provider API key. Set the environment variable named in apiKeyEnv.");
    }
    const raw = await postJson(
      `${trimTrailingSlash(this.baseUrl)}/chat/completions`,
      { model: this.model, messages, temperature: 0.2 },
      this.timeoutSeconds,
      {
        Authorization: `Bearer ${this.apiKey}`,
        ...optionalHeader("HTTP-Referer", process.env.BUTTERCLAW_SITE_URL),
        ...optionalHeader("X-Title", process.env.BUTTERCLAW_APP_NAME ?? "Butterclaw")
      }
    );
    const content = String(raw?.choices?.[0]?.message?.content ?? "");
    const usage = raw?.usage ?? {};
    return {
      content,
      promptTokens: Number(usage.prompt_tokens ?? 0),
      completionTokens: Number(usage.completion_tokens ?? 0),
      raw
    };
  }
}

export class OllamaProvider implements Provider {
  constructor(
    private readonly model: string,
    private readonly baseUrl: string,
    private readonly timeoutSeconds: number
  ) {}

  async complete(messages: Message[]): Promise<ProviderResponse> {
    const raw = await postJson(
      `${trimTrailingSlash(this.baseUrl)}/api/chat`,
      { model: this.model, messages, stream: false, options: { temperature: 0.2 } },
      this.timeoutSeconds,
      {}
    );
    return { content: String(raw?.message?.content ?? ""), raw };
  }
}

function mockToolResultAnswer(message: string): string {
  const result = message.match(/^Tool result for [^\n:]+:\r?\n(?:OK|ERROR): ([\s\S]*?)\r?\n\r?\nContinue\./i);
  const output = result?.[1]?.trim();
  const intro = "I checked the workspace and finished the requested step.";
  return output ? `${intro}\n\n${output}` : intro;
}

function optionalHeader(name: string, value: string | undefined): Record<string, string> {
  return value ? { [name]: value } : {};
}

async function postJson(
  url: string,
  payload: unknown,
  timeoutSeconds: number,
  headers: Record<string, string>
): Promise<any> {
  let response: Response;
  try {
    response = await fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json", Accept: "application/json", ...headers },
      body: JSON.stringify(payload),
      signal: AbortSignal.timeout(timeoutSeconds * 1000)
    });
  } catch (error) {
    if (isTimeoutError(error)) {
      throw new ProviderError(
        `Provider request timed out after ${timeoutSeconds}s. Try --request-timeout-seconds 180, a smaller task, or a faster model.`
      );
    }
    throw new ProviderError(`Provider request failed: ${error instanceof Error ? error.message : String(error)}`);
  }
  const text = await response.text();
  if (!response.ok) {
    throw new ProviderError(`Provider HTTP ${response.status}: ${text}`);
  }
  try {
    return JSON.parse(text);
  } catch {
    throw new ProviderError(`Provider returned non-JSON response: ${text.slice(0, 500)}`);
  }
}

function isTimeoutError(error: unknown): boolean {
  return error instanceof Error && (error.name === "TimeoutError" || error.name === "AbortError");
}

File: butterclaw/providers.py
from __future__ import annotations

import json
import os
import urllib.error
import urllib.request
from dataclasses import dataclass
from typing import Protocol, TypedDict

from butterclaw.config import ButterclawConfig


class Message(TypedDict):
    role: str
    content: str


@dataclass
class ProviderResponse:
    content: str
    prompt_tokens: int | None = None
    completion_tokens: int | None = None
    raw: dict[str, object] | None = None


class Provider(Protocol):
    def complete(self, messages: list[Message]) -> ProviderResponse:
        ...


class ProviderError(RuntimeError):
    pass


def estimate_tokens(text: str) -> int:
    return max(1, (len(text) + 3) // 4)


def build_provider(config: ButterclawConfig) -> Provider:
    if config.provider == "mock":
        return MockProvider()
    if config.provider == "ollama":
        return OllamaProvider(
            model=config.model,
            base_url=config.base_url or "http://localhost:11434",
            timeout=config.request_timeout_seconds,
        )
    if config.provider == "openai-compatible":
        return OpenAICompatibleProvider(
            model=config.model,
            base_url=config.base_url or "https://api.openai.com/v1",
            api_key=os.environ.get(config.api_key_env, ""),
            timeout=config.request_timeout_seconds,
        )
    raise ProviderError(f"Unknown provider: {config.provider}")


class MockProvider:
    """Deterministic provider used for smoke tests and offline demos."""

    def complete(self, messages: list[Message]) -> ProviderResponse:
        last = messages[-1]["content"].lower()
        if "tool result for" in last:
            return ProviderResponse(content="I checked the workspace and finished the requested step.")
        if "list" in last and ("file" in last or "workspace" in last):
            return ProviderResponse(content='{"tool": "list_dir", "args": {"path": "."}}')
        if "read" in last and "readme" in last:
            return ProviderResponse(content='{"tool": "read_file", "args": {"path": "README.md"}}')
        return ProviderResponse(
            content=(
                "Butterclaw mock provider is running. Switch to --provider ollama "
                "or --provider openai-compatible for real model reasoning."
            )
        )


class OpenAICompatibleProvider:
    def __init__(self, model: str, base_url: str, api_key: str, timeout: int = 60) -> None:
        self.model = model
        self.base_url = base_url.rstrip("/")
        self.api_key = api_key
        self.timeout = timeout

    def complete(self, messages: list[Message]) -> ProviderResponse:
        if not self.api_key:
            raise ProviderError("Missing API key. Set the configured api_key_env variable.")
        payload = {
            "model": self.model,
            "messages": messages,
            "temperature": 0.2,
        }
        raw = post_json(
            f"{self.base_url}/chat/completions",
            payload,
            timeout=self.timeout,
            headers={"Authorization": f"Bearer {self.api_key}"},
        )
        try:
            message = raw["choices"][0]["message"]["content"]  # type: ignore[index]
        except (KeyError, IndexError, TypeError) as exc:
            raise ProviderError(f"Unexpected OpenAI-compatible response: {raw}") from exc
        usage = raw.get("usage", {}) if isinstance(raw, dict) else {}
        return ProviderResponse(
            content=str(message or ""),
            prompt_tokens=int(usage.get("prompt_tokens", 0) or 0),
            completion_tokens=int(usage.get("completion_tokens", 0) or 0),
            raw=raw if isinstance(raw, dict) else None,
        )


class OllamaProvider:
    def __init__(self, model: str, base_url: str, timeout: int = 60) -> None:
        self.model = model
        self.base_url = base_url.rstrip("/")
        self.timeout = timeout

    def complete(self, messages: list[Message]) -> ProviderResponse:
        payload = {
            "model": self.model,
            "messages": messages,
            "stream": False,
            "options": {"temperature": 0.2},
        }
        raw = post_json(f"{self.base_url}/api/chat", payload, timeout=self.timeout, headers={})
        try:
            content = raw["message"]["content"]  # type: ignore[index]
        except (KeyError, TypeError) as exc:
            raise ProviderError(f"Unexpected Ollama response: {raw}") from exc
        return ProviderResponse(content=str(content or ""), raw=raw if isinstance(raw, dict) else None)


def post_json(
    url: str,
    payload: dict[str, object],
    timeout: int,
    headers: dict[str, str],
) -> dict[str, object]:
    body = json.dumps(payload).encode("utf-8")
    request = urllib.request.Request(
        url,
        data=body,
        headers={
            "Content-Type": "application/json",
            "Accept": "application/json",
            **headers,
        },
        method="POST",
    )
    try:
        with urllib.request.urlopen(request, timeout=timeout) as response:
            text = response.read().decode("utf-8")
    except urllib.error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise ProviderError(f"Provider HTTP {exc.code}: {detail}") from exc
    except urllib.error.URLError as exc:
        raise ProviderError(f"Provider request failed: {exc.reason}") from exc

    try:
        parsed = json.loads(text)
    except json.JSONDecodeError as exc:
        raise ProviderError(f"Provider returned non-JSON response: {text[:500]}") from exc
    if not isinstance(parsed, dict):
        raise ProviderError(f"Provider returned unexpected JSON: {parsed!r}")
    return parsed

